src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "drivers/time.h"
+#include "flight/failsafe.h"
+#include "flight/imu.h"
+#include "flight/position.h"
+#include "flight/gps_rescue.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "rx/rx.h"
+#include "fc/rc_controls.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 4);
+
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+
+    .maxVerticalVelocity = 30,
+);
+
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
```

Comment:
i would swap these lines so that you integrate first and then apply the i part (this is either euler forward or backward integration):
simplePid->integral += error * dt;
float iOut = simplePid->ki * simplePid->integral;
for the derivative you could only use d/dt(-measuredValue) ( instead of d/dt(targetValue - measuredValue)). so you cancel one zero in the system dynamics of each control loop (overshoot, only matters if setpoints changes anyway).